src/t8_forest/t8_forest_adapt.cxx

```diff
@@ -269,32 +279,144 @@ t8_forest_adapt (t8_forest_t forest)
+                                                                     el_considered - zz),
+                                      element_parent_compare);
+          if (!tscheme->t8_element_compare(element_parent_current, element_parent_compare)) {
+            is_family = 0;
+          }
+        }
+
+        /* Test 2: Check if elements in elements_from_copy get "eaten" by coarsing current element */
+        if (is_family) {
+          for (zz = 0; zz < num_children &&
+                       el_considered + (t8_locidx_t) zz - (t8_locidx_t) el_c < num_el_from; zz++) {
+            level = tscheme->t8_element_level(elements_from_copy[zz]);
+            /* Only elements with higher level then level of current element, can get eaten. */
+            if (level > level_current) {
+              tscheme->t8_element_copy(elements_from_copy[zz], element_parent_compare);
+              while(level > level_current-1) {
+                tscheme->t8_element_parent (element_parent_compare, element_parent_compare);
+                level = tscheme->t8_element_level(element_parent_compare);
+              }
+              if (!tscheme->t8_element_compare(element_parent_current, element_parent_compare)) {
+                is_family = 0;
+              }
+            }
+          }
+        }
+
+        /* Test 3: Check num_childrem elements before el_considered, if they will get eaten.
+         * Reason: current element could be on boarder
+         * [IL] Question: Is this required if el_c > 0
+         * */
+        if (is_family && el_considered > (t8_locidx_t) num_children) {
+          for (zz = 0; zz < num_children &&
+                       el_considered + (t8_locidx_t) zz - (t8_locidx_t) el_c < num_el_from; zz++) {
+              elements_from_copy[zz] = t8_element_array_index_locidx (telements_from,
+                                                                     el_considered + 
+                                                                     zz - num_children);
+          }
+          /* From here, it is the same test as test 2. 
+           * [IL] Question: Ca we do it in one step and is it wise? */
+          for (zz = 0; zz < num_children&&
+                       el_considered + (t8_locidx_t) zz - (t8_locidx_t) el_c < num_el_from; zz++) {
+            level = tscheme->t8_element_level(elements_from_copy[zz]);
+            if (level > level_current) {
+              tscheme->t8_element_copy(elements_from_copy[zz], element_parent_compare);
+              while(level > level_current-1) {
+                tscheme->t8_element_parent (element_parent_compare, element_parent_compare);
+                level = tscheme->t8_element_level(element_parent_compare);
+              }
+              if (!tscheme->t8_element_compare(element_parent_current, element_parent_compare)) {
+                is_family = 0;
+              }
+            }
+          } 
+        }
```

Comment:
These three test enlarge the already lenghty adapt-function. Can you pack them in a seperate function that is called by adapt? That way the code is probably better structured and more readable.